src/types.rs

```diff
@@ -764,3 +776,178 @@ impl LateLintPass for AbsurdExtremeComparisons {
+            (&FullInt::U(s), &FullInt::S(o)) => Self::cmp_s_u(o, s).reverse(),
+        })
+    }
+}
+impl Ord for FullInt {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).expect("partial_cmp for FullInt can never return None")
+    }
+}
+
+
+fn numeric_cast_precast_bounds<'a>(cx: &LateContext, expr: &'a Expr) -> Option<(FullInt, FullInt)> {
+    use rustc::ty::TypeVariants::{TyInt, TyUint};
+    use syntax::ast::UintTy;
+    use syntax::ast::IntTy;
+    use std::*;
+
+    if let ExprCast(ref cast_exp,_) = expr.node {
+        match cx.tcx.expr_ty(cast_exp).sty {
+            TyInt(int_ty) => Some(match int_ty {
+                IntTy::I8 => (FullInt::S(i8::min_value() as i64), FullInt::S(i8::max_value() as i64)),
+                IntTy::I16 => (FullInt::S(i16::min_value() as i64), FullInt::S(i16::max_value() as i64)),
+                IntTy::I32 => (FullInt::S(i32::min_value() as i64), FullInt::S(i32::max_value() as i64)),
+                IntTy::I64 => (FullInt::S(i64::min_value() as i64), FullInt::S(i64::max_value() as i64)),
+                IntTy::Is => (FullInt::S(isize::min_value() as i64), FullInt::S(isize::max_value() as i64)),
+            }),
+            TyUint(uint_ty) => Some(match uint_ty {
+                UintTy::U8 => (FullInt::U(u8::min_value() as u64), FullInt::U(u8::max_value() as u64)),
+                UintTy::U16 => (FullInt::U(u16::min_value() as u64), FullInt::U(u16::max_value() as u64)),
+                UintTy::U32 => (FullInt::U(u32::min_value() as u64), FullInt::U(u32::max_value() as u64)),
+                UintTy::U64 => (FullInt::U(u64::min_value() as u64), FullInt::U(u64::max_value() as u64)),
+                UintTy::Us => (FullInt::U(usize::min_value() as u64), FullInt::U(usize::max_value() as u64)),
+            }),
+            _ => None,
+        }
+    } else {
+        None
+    }
+}
+
+fn node_as_const_fullint(cx: &LateContext, expr: &Expr) -> Option<FullInt> {
+    use rustc::middle::const_eval::EvalHint::ExprTypeChecked;
+    use rustc_const_eval::*;
+
+    match const_eval::eval_const_expr_partial(cx.tcx, expr, ExprTypeChecked, None) {
+        Ok(val) => {
+            if let Integral(const_int) = val {
+                Some(match const_int {
+                    I8(x) => FullInt::S(x as i64),
+                    I16(x) => FullInt::S(x as i64),
+                    I32(x) => FullInt::S(x as i64),
+                    Isize(Is32(x)) => FullInt::S(x as i64),
+                    Isize(Is64(x)) |
+                    I64(x) => FullInt::S(x),
+                    InferSigned(x) => FullInt::S(x as i64),
+                    U8(x) => FullInt::U(x as u64),
+                    U16(x) => FullInt::U(x as u64),
+                    U32(x) => FullInt::U(x as u64),
+                    Usize(Us32(x)) => FullInt::U(x as u64),
+                    Usize(Us64(x)) |
+                    U64(x) => FullInt::U(x),
+                    Infer(x) => FullInt::U(x as u64),
```

Comment:
You could use the [`erase_type`](https://github.com/rust-lang/rust/blob/44a77f67696e855a9841eaf7c7bf7639131cfec6/src/librustc_const_eval/int.rs#L111-L112) method to only have two kind to handle here.